### PR TITLE
Remove useless whitespace in Websocket output

### LINKED ISSUE #9696: Reducing size of data packet of web socket
## Summary

Currently, the websocket push a well-format json text, which is not efficient and not necessary.

## Problem Definition

it increase usages of the bandwidth

## Proposal

I would like to suggest that we use compact json format to reduce the size of data packet.

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "query": "tm.event='NewBlock'",
    "data": {
      "type": "tendermint/event/NewBlock",
      "value": {
        "block": {
          "header": {
            "version": {
              "block": "11"
            },
            "chain_id": "cosmoshub-4",
            "height": "12798821",
            "time": "2022-11-11T03:13:43.729399725Z",
            "last_block_id": {
              "hash": "120A987F7F08B73662C5DD02E21D51042148216185AA7AC4150DB80C5DAB23C7",
              "parts": {
                "total": 1,
                "hash": "C46D66314E7ECB74614B0501433A864CA638C11155818AE15F2CF93AF59B8A71"
              }
            },
            "last_commit_hash": "5DC996F57B1B12E7107D5A271973315A750677B84D05B0799B2183B4D40941EC",
            "data_hash": "445049B9704FC63DA35838341BBAE494B8D71BDA1B6B652A61ABA0046F2D7597",
            "validators_hash": "7C016C1EEF692BCB44BBA640969286D6E96716FA878C8C67765CB0510A2161C2",
            "next_validators_hash": "906B9185EA6D07E5E59270DD4849A82BDAEE13A73DABA76826C656CC0ADA1BD2",
            "consensus_hash": "80364965B7C2CC9DE961C0998B47A7F93F1970077EB882E0ED1C3822408888C7",
            "app_hash": "3738062FB00F9107C750B6FC9CD931F2A1B8919BE19CCC58592F0CF62EC6ACD9",
            "last_results_hash": "BC45DF6713ACAF7CE3AB5B80C8C6BB8E4910774F9993A89DDE0F1DF9160A34C4",
            "evidence_hash": "E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855",
            "proposer_address": "D68EEC0D2E8248F1EC64CDB585EDB61ECA432BD8"
          },
          "data": {
               ...
          }
    }
  }
}
```
This block is about 280K,  

```json
{"jsonrpc":"2.0","id":1,"result":{"query":"tm.event='NewBlock'","data":{"type":"tendermint/event/NewBlock","value":{"block":{"header":{"version":{"block":"11"},"chain_id":"cosmoshub-4","height":"12798821","time":"2022-11-11T03:13:43.729399725Z","last_block_id":{"hash":"120A987F7F08B73662C5DD02E21D51042148216185AA7AC4150DB80C5DAB23C7","parts":{"total":1,"hash":"C46D66314E7ECB74614B0501433A864CA638C11155818AE15F2CF93AF59B8A71"}},"last_commit_hash":"5DC996F57B1B12E7107D5A271973315A750677B84D05B0799B2183B4D40941EC","data_hash":"445049B9704FC63DA35838341BBAE494B8D71BDA1B6B652A61ABA0046F2D7597","validators_hash":"7C016C1EEF692BCB44BBA640969286D6E96716FA878C8C67765CB0510A2161C2","next_validators_hash":"906B9185EA6D07E5E59270DD4849A82BDAEE13A73DABA76826C656CC0ADA1BD2","consensus_hash":"80364965B7C2CC9DE961C0998B47A7F93F1970077EB882E0ED1C3822408888C7","app_hash":"3738062FB00F9107C750B6FC9CD931F2A1B8919BE19CCC58592F0CF62EC6ACD9","last_results_hash":"BC45DF6713ACAF7CE3AB5B80C8C6BB8E4910774F9993A89DDE0F1DF9160A34C4","evidence_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","proposer_address":"D68EEC0D2E8248F1EC64CDB585EDB61ECA432BD8"},"data":{}}}}}
```
and this is 160k, we simply use compact mode by removing spaces.  

saved 1/3,

Users/developers can pretty it by themself

### PR DESCRIPTION
**Pull request by adizere**
Closes #9696 

---

#### PR checklist

- [x] Tests written/updated, or no tests needed
- [x] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [x] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

**Review by thanethomson — COMMENTED** [review]
Thanks for this @adizere! Could you please remove all prettified JSON RPC responses though, including the HTTP ones?

I think these are the only two remaining places:

https://github.com/tendermint/tendermint/blob/2b4436d1b4ed3c721587d9e2ce965dd38b661ca2/rpc/jsonrpc/server/http_server.go#L107

https://github.com/tendermint/tendermint/blob/2b4436d1b4ed3c721587d9e2ce965dd38b661ca2/rpc/jsonrpc/server/http_server.go#L143

**Comment by adizere** [comment]
With the new branch, the output now looks as follows

### For websocket output:

```bash
$ wscat -c ws://127.0.0.1:26657/websocket
Connected (press CTRL+C to quit)
> { "jsonrpc": "2.0", "method": "subscribe", "id": 0, "params": { "query": "tm.event='NewBlock'" } }
< {"jsonrpc":"2.0","id":0,"result":{}}
< {"jsonrpc":"2.0","id":0,"result":{"query":"tm.event='NewBlock'","data":{"type":"tendermint/event/NewBlock","value":{"block":{"header":{"version":{"block":"11","app":"1"},"chain_id":"test-chain-gjnVQx","height":"487","time":"2022-11-18T13:46:34.893604324Z","last_block_id":{"hash":"076C580AA40A471DE818940A0C78A0F4ADF69A92455AE60904366AD7F6CB0CDA","parts":{"total":1,"hash":"0FCF10C92D65D08471FED7EBC8697ACCAC2DE25A31FBD2CEB8565A2B514442A8"}},"last_commit_hash":"29F2B040E79C9CEBE7794FF800C9BEAC1F65D47963ED985C8361035EC76E820C","data_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","validators_hash":"C8B9A24D7717676A0BDDB1C17F9499D9A50E1D756420A791FD93948010629973","next_validators_hash":"C8B9A24D7717676A0BDDB1C17F9499D9A50E1D756420A791FD93948010629973","consensus_hash":"048091BC7DDC283F77BFBF91D73C44DA58C3DF8A9CBC867405D8B7F3DAADA22F","app_hash":"0000000000000000","last_results_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","evidence_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","proposer_address":"00F9A04401951975AB4445209672640EF0E187C8"},"data":{"txs":[]},"evidence":{"evidence":[]},"last_commit":{"height":"486","round":0,"block_id":{"hash":"076C580AA40A471DE818940A0C78A0F4ADF69A92455AE60904366AD7F6CB0CDA","parts":{"total":1,"hash":"0FCF10C92D65D08471FED7EBC8697ACCAC2DE25A31FBD2CEB8565A2B514442A8"}},"signatures":[{"block_id_flag":2,"validator_address":"00F9A04401951975AB4445209672640EF0E187C8","timestamp":"2022-11-18T13:46:34.893604324Z","signature":"nlPcmDO+rlXN8alPxXGLDxllRHBaH0US37NgCJCzR2Jj4zM25AHsMHN4bd/VSUzRBfiuRobC/rSe1A5zcEACBA=="}]}},"result_begin_block":{},"result_end_block":{"validator_updates":null}}},"events":{"tm.event":["NewBlock"]}}}
< {"jsonrpc":"2.0","id":0,"result":{"query":"tm.event='NewBlock'","data":{"type":"tendermint/event/NewBlock","value":{"block":{"header":{"version":{"block":"11","app":"1"},"chain_id":"test-chain-gjnVQx","height":"488","time":"2022-11-18T13:46:35.906382967Z","last_block_id":{"hash":"755DB59E9DFBF1F8A20948515DF4863D17FE46E51A1724A73994AB9A2B6F3CE1","parts":{"total":1,"hash":"CCFC30A2BEE2D0A2AF4CB397FB5BBDFC1FA69C3342509769CAB645627E6C4DA6"}},"last_commit_hash":"F705495755751E3E55BE40CA9E71BF26485AFB4004D07B21C20AE78D4ECA7973","data_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","validators_hash":"C8B9A24D7717676A0BDDB1C17F9499D9A50E1D756420A791FD93948010629973","next_validators_hash":"C8B9A24D7717676A0BDDB1C17F9499D9A50E1D756420A791FD93948010629973","consensus_hash":"048091BC7DDC283F77BFBF91D73C44DA58C3DF8A9CBC867405D8B7F3DAADA22F","app_hash":"0000000000000000","last_results_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","evidence_hash":"E3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B934CA495991B7852B855","proposer_address":"00F9A04401951975AB4445209672640EF0E187C8"},"data":{"txs":[]},"evidence":{"evidence":[]},"last_commit":{"height":"487","round":0,"block_id":{"hash":"755DB59E9DFBF1F8A20948515DF4863D17FE46E51A1724A73994AB9A2B6F3CE1","parts":{"total":1,"hash":"CCFC30A2BEE2D0A2AF4CB397FB5BBDFC1FA69C3342509769CAB645627E6C4DA6"}},"signatures":[{"block_id_flag":2,"validator_address":"00F9A04401951975AB4445209672640EF0E187C8","timestamp":"2022-11-18T13:46:35.906382967Z","signature":"NNKamS/P1dnRpBeLAgq/A7zHaCnuZvLxk3VASWFnTxF8usWVMFEO3ik43YHnmk8EFTVAi9mLY+tanmsRL+E9AQ=="}]}},"result_begin_block":{},"result_end_block":{"validator_updates":null}}},"events":{"tm.event":["NewBlock"]}}}
```

It used to be:

```bash
> { "jsonrpc": "2.0", "method": "subscribe", "id": 0, "params": { "query": "tm.event='NewBlock'" } }
< {
  "jsonrpc": "2.0",
  "id": 0,
  "result": {}
}
< {
  "jsonrpc": "2.0",
  "id": 0,
  "result": {
    "query": "tm.event='NewBlock'",
    "data": {
      "type": "tendermint/event/NewBlock",
      "value": {
        "block": {
          "header": {
            "version": {
              "block": "11",
              "app": "1"
....
```


### For RPC output:

It is now:

```bash
$ curl http://localhost:26657/status
{"jsonrpc":"2.0","id":-1,"result":{"node_info":{"protocol_version":{"p2p":"8","block":"11","app":"1"},"id":"0c47498ed3cfe332e60e6dd07597509064e92a82","listen_addr":"tcp://0.0.0.0:26656","network":"test-chain-gjnVQx","version":"0.38.0-dev","channels":"40202122233038606100","moniker":"hermes-bastion-tor1","other":{"tx_index":"on","rpc_address":"tcp://127.0.0.1:26657"}},"sync_info":{"latest_block_hash":"1A17BF7FABCDA1E0A1D3C7FC8F8854A831A53C718ABBDBCB8F2FAA478AB3F0A3","latest_app_hash":"0000000000000000","latest_block_height":"467","latest_block_time":"2022-11-18T13:46:14.581563537Z","earliest_block_hash":"32619EBAA20611BB1D0D5E731E582A62AA7DD5027F831D50E98C819189531DB8","earliest_app_hash":"","earliest_block_height":"1","earliest_block_time":"2022-11-18T13:32:21.415141857Z","catching_up":false},"validator_info":{"address":"00F9A04401951975AB4445209672640EF0E187C8","pub_key":{"type":"tendermint/PubKeyEd25519","value":"WMRogr4A8f7gkL+rGG9GVP55MycgsiKp25rsl7YlTd0="},"voting_power":"10"}}}
```

It used to be:

```bash
$ curl http://localhost:26657/status
{
  "jsonrpc": "2.0",
  "id": -1,
  "result": {
    "node_info": {
      "protocol_version": {
        "p2p": "8",
        "block": "11",
        "app": "1"
...
```

**Comment by adizere** [comment]
Something to consider: could this change be breaking for some users?

I imagine some users of the binary might have scripts doing `$ curl http://localhost:26657/status | fgrep 'catching_up'` (or similar). If such a script parses & expects a `"catching_up": false` or `"catching_up": true` on a single line, this change will be breaking.

**Review by thanethomson — COMMENTED** [review]
Awesome, thanks! I see there is one test that's failing that looks like it's due to these changes: https://github.com/tendermint/tendermint/actions/runs/3497400805/jobs/5856450702

It looks like the tests are matching on raw HTTP responses in https://github.com/tendermint/tendermint/blob/main/rpc/jsonrpc/server/http_server_test.go

**Comment by adizere** [comment]
Thanks Thane! Tests seem to pass now.

It was not clear if the changelog pending entry should go into `0.37` or `0.38` section of `CHANGELOG_PENDING.md`. I assumed it should be `0.38` -- 90bee8d36 glad to fix it if that's not right.

**Review by liangping — APPROVED** [review]
Yeah. it's a great improvement.

**Comment by thanethomson** [comment]
> It was not clear if the changelog pending entry should go into 0.37 or 0.38 section of CHANGELOG_PENDING.md. I assumed it should be 0.38 -- https://github.com/tendermint/tendermint/commit/90bee8d361b8c97c96f06426af89473780b2d9cd glad to fix it if that's not right.

In this case, since we're backporting it and it'll land up in v0.34 and v0.37.1, it probably makes sense to add it to those release branches' changelogs. Users can safely assume that in v0.38 it'll be present too, and so it wouldn't really be necessary to add it to the v0.38 changelog.

I'll make the necessary changes.

File: CHANGELOG_PENDING.md [hunk 1]
```diff
@@ -34,6 +34,7 @@
 - [p2p/pex] \#6509 Improve addrBook.hash performance (@cuonglm)
 - [crypto/merkle] \#6443 & \#6513 Improve HashAlternatives performance (@cuonglm, @marbar3778)
 - [rpc] \#9650 Enable caching of RPC responses (@JayT106)
+- [rpc] \#9720 Remove useless whitespace in RPC output (@adizere, @thanethomson)
```
Comment (thanethomson) [inline CHANGELOG_PENDING.md:37]:
```suggestion
```

We can add this to the v0.37 and v0.34 changelogs.

**Pull request merged**